In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [2]:
import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        20:23:38   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

4.  Explain the apparent contradiction between the factual statements regarding the fit below that "the model only explains 17.6% of the variability in the data" while at the same time "many of the coefficients are larger than 10 while having strong or very strong evidence against the null hypothesis of 'no effect'"

The contradiction between the model’s low R-squared value (17.6%) and the strong significance of certain coefficients can be explained by understanding that R-squared and p-values measure different things. An R-squared of 0.176 means the model explains only 17.6% of the variation in HP, so there are likely other factors affecting HP that the model doesn’t capture. However, the low p-values and large coefficients for certain predictors show that these predictors still have an important effect on HP, even if the model as a whole doesn’t explain all of HP’s variability.

When creating this model, the Generation variable was treated as a categorical variable to capture the unique effects that each generation may have on HP. An interaction term between Sp. Def and Generation was also added, allowing Sp. Def to have a different effect on HP depending on the generation. This means Sp. Def might have a stronger impact on HP for certain generations than others, which the interaction term helps to capture. For example, the impact of Sp. Def on HP could be much stronger in certain generations, and this setup lets the model account for those differences.

In this way, R-squared and p-values complement each other: R-squared measures the overall fit of the model, while p-values tell us how important each predictor is. Even with a low R-squared, low p-values and large coefficients for certain predictors mean those predictors still significantly influence HP.

7. Discuss with a ChatBot the rationale and principles by which model5_linear_form is extended and developed from model3_fit and model4_fit; model6_linear_form is extended and developed from model5_linear_form; and model7_linear_form is extended and developed from model6_linear_form; then, explain this breifly and consisely in your own words

The phases from model3_fit to model7_linear_form thoughtfully build on each other to improve prediction accuracy, interpretability, and model stability. From model5_linear_form, the model includes Attack, Defense, Speed, Generation, and Type as predictors. Adding these variables captures more HP impacts, increasing the R-squared value by covering more data variance. Generation and Type help the algorithm discover distinctions across Pokémon kinds and generations, making predictions more adaptive and useful across Pokémon categories. Model6_linear_form simplifies and focuses the model. This model uses just model5's most important predictors—Attack, Speed, Sp. Def, and Sp. Atk—and adds binary indicators to indicate type and generation characteristics. To identify Normal, Water, and Generation 2 or 5 Pokémon, markers are added. This careful use of indicators keeps the model focused on the most important HP predictors while capturing unique qualities without over-complicating the structure. This update clarifies how each factor predicts HP, making it simpler to validate the model's assumptions and more practical for users who need to comprehend its predictions. Model7_linear_form introduces interaction terms between important predictors like Attack, Speed, Sp. Def, and Sp. Atk to show how they affect HP differently. Interaction terms enable the model to account for complicated interactions where one variable's influence on HP depends on another variable's level, yielding a more nuanced forecast. This accuracy is useful, but it complicates the model and makes it difficult to assess each predictor's independent influence. Model7_linear_form_CS centers and scales continuous predictors to handle complexity and multicollinearity. This tweak stabilizes the model by lowering multicollinearity and standardizing coefficients for better interpretation. Centering and scaling make predictions consistent and the model more adaptable to data changes by adding interaction terms without compromising stability.

Question 7 chat bot session: https://chatgpt.com/share/67368f69-e2ec-8009-a664-9809ca7d9407

9. Work with a ChatBot to understand the meaning of the illustration below; and, explain this in your own words

Initially, `model7_fit` has a greater "in sample" R-squared and it suggests a strong fit and significant training data variance. However, complexity increases overfitting risk. Overfitting causes a model to learn characteristics and noise from training data that don't apply to new data, reducing prediction accuracy. Data may deviate greatly from the original dataset in real-world applications, which is concerning. While model7_fit may perform well in a controlled training scenario, its performance may decline with diverse data.

In comparison, `model6_fit` has a simpler structure and reduces complexity, resulting in greater performance on new data without overfitting. Because of its stability over numerous datasets, Model 6_fit excels in real-world applications needing reliable predictions. Model6_fit provides more balanced and generally applicable solutions than `model7_fit`, yielding accurate forecasts for new data.

In contrast to `model7_fit`, `model6_fit` has simpler interactions and is easier to comprehend. For domains that need outcomes insights, Model6_fit is better at studying factor effects. Although `model7_fit` may detect tiny correlations, its complex interactions make interpretation difficult. High interaction rates may induce multicollinearity, where tightly related variables affect each other's coefficients, reducing model stability and accuracy. Multicollinearity may cause large model output variations from small data changes, lowering model prediction confidence. Selecting `model6_fit` yields clearer, more interpretable results.

Modeling requires balancing accuracy, interpretability, and generalization. Overfitting may lead to poor performance of complex models on new data, despite high accuracy on training data. This balance is presented in Python code using Generations 1–5 for training and Generation 6 for testing and it allows us to compare model performance on new data. Results show that `model6_fit` is more accurate across different data subsets, but `model7_fit` can struggle to generalize to new data outside the training set owing to its complex structure.

To get dependable and intelligible findings, consider using `model6_fit` instead of `model7_fit`, which may have greater prediction accuracy on training data owing to its complexity and interaction terms. Python code can assess model strengths and weaknesses, indicating `model6_fit` strikes a compromise between accuracy and stability across datasets. A balanced approach is needed to create models that perform in real-world circumstances with data unpredictability and clear interpretation.

Question 9 chat bot session: https://chatgpt.com/share/67369f36-98b4-8009-97a4-1f3eb215867e